SMLHelper/Patchers/CraftDataPatcher_Subnautica.cs

```diff
@@ -22,76 +26,314 @@ internal static void AddToCustomTechData(TechType techType, ITechData techData)
+               prefix: new HarmonyMethod(AccessTools.Method(typeof(CraftDataPatcher), nameof(GetBuildablePrefix))));
+
+        }
+
+        private static void PatchModPrefabs()
+        {
+            Dictionary<TechType, string> techMapping = CraftData.techMapping;
+            Dictionary<string, TechType> entClassTechTable = CraftData.entClassTechTable;
+            foreach (ModPrefab prefab in ModPrefab.Prefabs)
+            {
+                techMapping[prefab.TechType] = prefab.ClassID;
+                entClassTechTable[prefab.ClassID] = prefab.TechType;
+            }
+        }
+
+        private static void GetHarvestOutputPrefix(TechType techType)
+        {
+            if (CraftData.harvestOutputList.ContainsKey(techType) && CustomHarvestOutputList.ContainsKey(techType))
+            {
+                if(CraftData.harvestOutputList[techType] != CustomHarvestOutputList[techType])
+                    CraftData.harvestOutputList[techType] = CustomHarvestOutputList[techType];
+            }
+            else if (CustomHarvestOutputList.ContainsKey(techType))
+            {
+                PatchUtils.PatchDictionary(CraftData.harvestOutputList, CustomHarvestOutputList);
+            }
+        }
+
+        private static void GetHarvestTypePrefix(TechType techType)
+        {
+            if (CraftData.harvestTypeList.ContainsKey(techType) && CustomHarvestTypeList.ContainsKey(techType))
+            {
+                if (CraftData.harvestTypeList[techType] != CustomHarvestTypeList[techType])
+                    CraftData.harvestTypeList[techType] = CustomHarvestTypeList[techType];
+            }
+            else if (CustomHarvestTypeList.ContainsKey(techType))
+            {
+                PatchUtils.PatchDictionary(CraftData.harvestTypeList, CustomHarvestTypeList);
+            }
         }
 
-        private static void AddCustomTechDataToOriginalDictionary()
+        private static void GetFinalCutBonusPrefix(TechType techType)
         {
+            if (CraftData.harvestFinalCutBonusList.ContainsKey(techType) && CustomFinalCutBonusList.ContainsKey(techType))
+            {
+                if (CraftData.harvestFinalCutBonusList[techType] != CustomFinalCutBonusList[techType])
+                    CraftData.harvestFinalCutBonusList[techType] = CustomFinalCutBonusList[techType];
+            }
+            else if (CustomFinalCutBonusList.ContainsKey(techType))
+            {
+                PatchUtils.PatchDictionary(CraftData.harvestFinalCutBonusList, CustomFinalCutBonusList);
+            }
+        }
+
+        private static void GetItemSizePrefix(TechType techType)
+        {
+            if (CraftData.itemSizes.ContainsKey(techType) && CustomItemSizes.ContainsKey(techType))
+            {
+                if ((CraftData.itemSizes[techType].x != CustomItemSizes[techType].x || CraftData.itemSizes[techType].y != CustomItemSizes[techType].y))
+                    CraftData.itemSizes[techType] = CustomItemSizes[techType];
```

Comment:
Throughout these patching methods, I see the same pattern used several times
```CS
if (CraftData.somedictionary.ContainsKey(techType))
{
    /*do something with*/ CraftData.somedictionary[techType]
}
```

This pattern results in doing the key-lookup twice.  
Instead, use the `TryGetValue` method from `IDictionary` whenever possible.

```CS
if (CraftData.somedictionary.TryGetValue(techType, out var value))
{
    /*do something with*/ value
}
```

Reply:
looks like that would work nicely here. forgot that the try get value was only patched in the BZ side so i can use it here without causing an infinite loop.